JuliaStream.jl/src/AMDGPUStream.jl

```diff
@@ -0,0 +1,178 @@
+# AMDGPU.jl doesn't support CPU agents, so this isn't a feature-complete ROCmStream, only AMD GPUs
+include("Stream.jl")
+using AMDGPU
+
+const ROCData = StreamData{T,ROCArray{T}} where {T}
+const TBSize = 1024::Int
+const DotBlocks = 256::Int
+
+# AMDGPU.agents()'s internal iteration order isn't stable
+function gpu_agents_in_repr_order()
+  # XXX if we select anything other than :gpu, we get 
+  # HSA_STATUS_ERROR_INVALID_AGENT on the first kernel submission
+  sort(AMDGPU.get_agents(:gpu), by = repr)
+end
+
+function devices()
+  try
+    map(repr, gpu_agents_in_repr_order())
+  catch
+    # probably unsupported
+    []
+  end
+end
+
+function gridsize(data::ROCData{T})::Int where {T}
+  return data.size
+end
+
+function make_stream(
+  arraysize::Int,
+  scalar::T,
+  device::Int,
+  silent::Bool,
+)::ROCData{T} where {T}
+
+  if arraysize % TBSize != 0
+    error("arraysize ($(arraysize)) must be divisible by $(TBSize)!")
+  end
+
+  # XXX AMDGPU doesn't expose an API for setting the default like CUDA.device!()
+  # but AMDGPU.get_default_agent returns DEFAULT_AGENT so we can do it by hand
+  AMDGPU.DEFAULT_AGENT[] = gpu_agents_in_repr_order()[device]
+
+  data = ROCData{T}(
+    ROCArray{T}(undef, arraysize),
+    ROCArray{T}(undef, arraysize),
+    ROCArray{T}(undef, arraysize),
+    scalar,
+    arraysize,
+  )
+  selected = AMDGPU.get_default_agent()
+  if !silent
+    println("Using GPU HSA device: $(AMDGPU.get_name(selected)) ($(repr(selected)))")
+    println("Kernel parameters   : <<<$(gridsize(data)),$(TBSize)>>>")
+  end
+  return data
+end
+
+function hard_wait(kernel)
+  # soft wait causes HSA_REFCOUNT overflow issues
+  AMDGPU.wait(kernel, soft = false)
+end
+
+function init_arrays!(data::ROCData{T}, init::Tuple{T,T,T}) where {T}
+  AMDGPU.fill!(data.a, init[1])
+  AMDGPU.fill!(data.b, init[2])
+  AMDGPU.fill!(data.c, init[3])
+end
+
+function copy!(data::ROCData{T}) where {T}
+  function kernel(a, c)
+    i = (workgroupIdx().x - 1) * workgroupDim().x + workitemIdx().x
+    @inbounds c[i] = a[i]
+    return
+  end
+  hard_wait(@roc groupsize = TBSize gridsize = gridsize(data) kernel(data.a, data.c))
+end
+
+function mul!(data::ROCData{T}) where {T}
+  function kernel(b, c, scalar)
+    i = (workgroupIdx().x - 1) * workgroupDim().x + workitemIdx().x
```

Comment:
This looks weird. Why is the workgroup Idx -1 but not the workitem Idx? If workgroups are numbered from 1 (or 2) but work items are numbers from 0 (or 1), it's worth adding a comment.

Reply:
See comment on the CUDA one. Will add a comment in code.

Reply:
Done